src/arsenic/services.py

```diff
@@ -36,16 +40,18 @@ async def subprocess_based_service(
         session = ClientSession()
         closers.append(session.close)
         count = 0
-        while True:
-            try:
+
+        async def wait_service():
+            # Wait for service with exponential back-off
+            for i in range(-10, 0):
                 if await tasked(check_service_status(session, service_url)):
-                    break
-            except:
-                # TODO: make this better
-                count += 1
-                if count > 30:
-                    raise Exception("not starting?")
-                await asyncio.sleep(0.5)
+                    return
+                await asyncio.sleep(start_timeout * 2 ** i)
+
+        try:
+            await asyncio.wait_for(wait_service(), timeout=start_timeout)
+        except asyncio.TimeoutError:
+            raise ArsenicError("not starting?")
```

Comment:
I wonder if the additional timeout handling is needed...
on one hand, if `tasked()` took even a few seconds (possible if the HTTP request is somehow slow), the result would be additional tens of seconds of waiting, notably longer than `start_timeout`.
on the other, dealing with cancellation is hard...
@ojii what do you think?

Reply:
If you're expecting some kind of special delay because you've got an "interesting" environment, then you've probably set start_timeout to something sensible to account for that.
And if you're not and you blow the timeout budget on a few really slow service up checks, then you've probably got bigger problems :worried: